### ISSUE TITLE
macOS: Build a native app image that works without Docker

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#3627]
**As an Apple macOS user, I would like to install PhotoPrism without Docker so that it is easier to set up and has better performance.**

Thanks to your support, we finally have a Mac mini with the new M2 CPU, so we can start working on native packages for Apple users. Note that this will take a while. As soon as there is a testable version, we will announce it here and on [Mastodon](https://floss.social/@photoprism).

![mac-mini](https://github.com/photoprism/photoprism/assets/301686/bd6a8e3d-b615-488e-a4ed-98cc3b498f2b)

**Comment by headphonia** [comment]
Do you have any updates on the native macOS app? Thanks.

**Comment by lastzero** [comment]
@headphonia While it is possible to create native macOS binaries for the previously used Intel CPU architecture, we still lack suitable TensorFlow binaries that are compatible with the latest macOS on Apple Silicon:
- https://dl.photoprism.app/tensorflow/osx/

This required using a rather old C compiler that does not support Apple Silicon simply because it is too old. I would love to give it a try and find a way to make it work, but first need to complete work on 2FA and some other issues for our next release:
- https://docs.photoprism.app/release-notes/#development-preview
- https://github.com/orgs/photoprism/projects/5

Any help with this would of course be much appreciated! 💎

**Comment by ChR-iSz** [comment]
Hi,

[See here](https://stackoverflow.com/a/77630557)

**Comment by curves-dev** [comment]
wow! i want test this1

**Comment by dimatakoy** [comment]
Leets go? ;-D

**Comment by ChR-iSz** [comment]
😊 I will create a basic Application next time in SwiftUI. But it will be something very simple that meets my needs. If there is something presentable, I will be happy to publish the project here. I use an iMac with M2 chip.

**Comment by TaiPhamD** [comment]
This would be a big undertaking but would it be worth it to try to move away from tensorflow and use PyTorch (libtorchvision c++ api?) so that we can build it for apple devices and possibly better support with Pytorch. Torch vision has many pre-trained models: https://pytorch.org/vision/main/models.html in addition to support MPS kernel.

**Comment by gkatenin** [comment]
Hi,

Here's how to build the native Apple silicon app:

1. Get libtensorflow [here](https://storage.googleapis.com/tensorflow/versions/2.18.0/libtensorflow-cpu-darwin-arm64.tar.gz)

- Copy the library somewhere reachable in the lib path (/opt/local/lib in my case)

3. Clone photoprism:
- Edit scripts/build.sh , so that it includes the path to the library :
- % diff -w scripts/build.sh ~/Soft/photoprism/scripts/build.sh
23c23
<   BUILD_CMD=("$GO_BIN" build -tags="debug,develop" -ldflags "-X main.version=${BUILD_ID}-DEVELOP **-r /opt/local/lib**" -o "${BUILD_BIN}" cmd/photoprism/photoprism.go)
---
>   BUILD_CMD=("$GO_BIN" build -tags="debug,develop" -ldflags "-X main.version=${BUILD_ID}-DEVELOP" -o "${BUILD_BIN}" cmd/photoprism/photoprism.go)

3. Build the app :

- mkdir build
- % CGO_CPPFLAGS=-I/opt/local/include make build
rm -f photoprism
realpath: ./assets/static/build: No such file or directory
scripts/build.sh develop photoprism
Building PhotoPrism 250413-f09aa7749-Darwin-ARM64 (develop)...
=> compiling "photoprism" with "go version go1.24.2 darwin/arm64"
=> go build -tags=debug,develop -ldflags -X main.version=250413-f09aa7749-Darwin-ARM64-DEVELOP -r /opt/local/lib -o photoprism cmd/photoprism/photoprism.go
 80M	photoprism
Done.

5. Profit :

-  % ./photoprism version                    
250413-f09aa7749-Darwin-ARM64-DEVELOP

Hope this helps!

**Comment by lastzero** [comment]
@gkatenin Thanks for sharing! That's great, but of course, compiling from source isn't for everyone :)

I wish we could set aside time to work on an official macOS build. For now, though, we need to focus on Batch Edit and AI enhancements. We'll look into this as soon as possible.

If you'd like to help us create an installable bundle to offer for download or through the App Store, feel free to reach out to us via email or chat. 💬

**Comment by chrisbansart** [comment]
Hi,
I took the time to create a script to build a .pkg installer (based on the latest version of PhotoPrism with TensorFlow and ONNX) to simplify the installation of a PhotoPrism server on Mac—all while following macOS best practices.
Thanks to it you are be able to build with one command a .pkg installer for mac for each new photoprism release.

It works like a charm! The biggest challenge was gathering and integrating the libraries into the .pkg, but the result seems solid.
On top of that, I developed a Swift app to control the server effortlessly. I’ll push it to a repo after a few final tweaks (use right places to put photoprism folders on a mac corresponding to good practices on MacOS.
In the meantime, here’s a sneak peek to tease the product 😉:


<img width="895" height="608" alt="Image" src="https://github.com/user-attachments/assets/6f589588-ea29-4c6b-b650-6a23329c3485" />

Happy to help and share.
Chris

**Comment by lastzero** [comment]
Thanks for sharing! That's awesome. 👌

**Comment by chrisbansart** [comment]
Hi,
The repo is here : https://github.com/chrisbansart/macOS-Packager-for-PhotoPrism
Let me know if it works on your mac. I suppose we have to compile on each os x version to be sure the final PhotoPrism server is compiled for the right macOS version. Due to that, you will have probably distribute on .pkg (600Mo approx) per macOS X major version. On my side I've used the package on MacOS Sequoia 15 and it works fine.
I've also add some features to the launcher (In Preferences menu), but several tests has to be done before use it securily in production.
Thanks
Chris

**Comment by omerdduran** [comment]
Hi Chris,

I tested it on my MacBook (M2, macOS 26.1), and it works very well. The only issue I noticed is that when installing via the .pkg file, the PhotoPrism application isn’t automatically moved to the Applications folder — I have to move it manually. For now, I’m using this command:

sudo cp -R ~/GitHub/macOS-Packager-for-PhotoPrism/temp/pkgroot/Applications/PhotoPrism.app /Applications/


Other than that, everything seems to run smoothly.
Thanks again for your work on this!

**Comment by chrisbansart** [comment]
Hi @omerdduran, Thanks for your feedback. Has the .pkg installer asked for an admin password? Normally, it uses it to install in Application Folder : that's the good practice for .pkg installer. 
Could you test in a terminal the 2 following commandes  and copy - paste the result here: 
pkgutil --files app.photoprism.server

and 

show --predicate 'subsystem == "com.apple.installer"' --last 48h

Thanks.
PS : At the end, there are two good practices to install a macOS App in /Applications. The .pkg and the .dmg which displays two icons (PhotoPrism and Applications) and an arrow to visually request user to move PhotoPrism to the Applications Folder. Personnaly, I prefer the second, because it's more transparent and it show that there are no hidden files installed in the system. (All libraries are embed in the Photoprism app).

**Comment by chrisbansart** [comment]
I've updated the repo https://github.com/chrisbansart/macOS-Packager-for-PhotoPrism to build a .dmg "installer" which provides a better User experience.

**Comment by chrisbansart** [comment]
The solution seems robust, and the .dmg format is definitely the best option for distribution. However, I kept both options in the bash script primarily because it’s aimed at developers rather than end users. On that note, would you like to include these installers on the official website?

The next challenge for macOS users will be implementing a migration tool for Photos.app. I’ve explored several existing methods for migrating from Photos.app to PhotoPrism, but I’ve noticed that all of them are either unintuitive or result in data loss. For example, users must choose between original or edited photos—not both—and albums or tags aren’t migrated. This becomes a significant hurdle for macOS users with large libraries. Improving this process could make migration much smoother. I've imagined starting by wraping the [osxphotos](https://github.com/RhetTbull/osxphotos) in a simple swift App where you can set several parameters in order migrate photos (orignals, edited ones, tages, albums...) from Photos.app library to Photoprism folder. What do you think about that, any advice ? 

Chris

**Comment by graciousgrey** [comment]
Thank you very much for sharing! 

@lastzero will take a closer  look at it as soon as our time allows. In the meantime, you are welcome to open a PR to add your repo to our contribution collection so others can discover it sooner: https://github.com/photoprism/photoprism-contrib :)

**Comment by lastzero** [comment]
@chrisbansart We are definitely interested in integrating this into our CI pipeline to provide Mac users with a native build. That's why we opened this issue; we just never had enough time to move forward with it on our own, so your help is much appreciated! 🤗 

Since I still have many other tasks to take care of before the holidays, I may not be able to provide feedback until early next year. However, I'll monitor the comments on this issue, and you are welcome to reach out to us at any time with questions, also via email or chat on Element/Matrix.

**Comment by chrisbansart** [comment]
Noticed ! Meanwhile, I'll go further into the MacOS Photos.app to Photoprism migration tool ;-)

**Comment by ChR-iSz** [comment]
It's great that things are moving forward here. If I can help in any way or if you need to divide up the work, just let me know. I'm a full-stack developer...

**Comment by lvzhaobing** [comment]
I would like to ask if the final built DMG package includes and correctly handles the TensorFlow C libraries.  Also, what else do I need to do if I want to use TensorFlow with Core ML acceleration support on the Apple M4 platform?

**Comment by chrisbansart** [comment]
Yes, the final DMG package includes the TensorFlow C libraries required by PhotoPrism. They are bundled in the DMG and correctly linked at runtime, so no additional TensorFlow installation is needed for standard use.

As far as I know, on Apple Silicon, TensorFlow uses its own optimized kernels. Core ML acceleration is not used automatically by PhotoPrism.
If Core ML acceleration is desired, this would require:
	•	   Rebuilding TensorFlow Lite with the Core ML delegate enabled
	•	   Modifying PhotoPrism to load and use that delegate explicitly

On top of that, we still need to verify whether Core ML with TensorFlow actually delivers meaningful performance improvements; at this stage, it’s difficult to say...

In short: the DMG already includes everything required for PhotoPrism’s TensorFlow/TFLite usage, and Core ML acceleration is not supported out of the box on macOS.
Do you plan specific usage that needs CoreML customisation ?
Chris

**Comment by lvzhaobing** [comment]
No problem, I was just curious about the current state of hardware-accelerated TensorFlow support on the Apple M-series silicon platform.
Thank you for your answer and your great work!

**Comment by chrisbansart** [comment]
> Thank you very much for sharing!
> 
> [@lastzero](https://github.com/lastzero) will take a closer look at it as soon as our time allows. In the meantime, you are welcome to open a PR to add your repo to our contribution collection so others can discover it sooner: https://github.com/photoprism/photoprism-contrib :)

Hi @graciousgrey, I cannot open a PR on https://github.com/photoprism/photoprism-contrib to add my repo. Could tell me how to achieve that ?
Thanks, Chris
